### DISCUSSION TITLE
Can't call an S3 post from an API route when deployed on Vercel - but works locally and on other servers?

### DISCUSSION THREAD

**Discussion by rauchp** [vercel/next.js#21295] [Help] [answered]
Hey guys, so the problem I'm having is with uploading an image to a public S3 bucket from my `api/uploadImage` route. 

On my localhost, everything works perfectly. On the deployed version, I’m getting `AccessDenied` when I try to upload through the S3 API. I've tried getting pre-signed URL’s and just POST’ing my images to the bucket, my POST's end up `forbidden` regardless.

From what I’ve gathered, the problem is posting from the Vercel domain to AWS. On my localhost, I can use s3.upload / POST with a pre-signed URL. Whereas on Vercel’s servers I get 403 when I try to do the same. I was able to upload images when I `yarn build && yarn start` on a Digital Ocean droplet, so I think it has something to do with Vercel’s network. (But not completely sure.)

Initially, I thought this was an AWS permissions problem.  I set my bucket to the most public CORS and bucket policies. I still only see this problem on Vercel’s server, not my localhost or my Digital Ocean server. 

I can call GET functions from the AWS JS SDK just fine, so it’s not IAM role / account authorizations or environment variables. 

Has anyone seen something like this? Thanks!

**Comment by DopamineDriven** [comment]
This is interesting because it is an xyz.cloudfront.net domain for S3, right? I use headless wordpress via Amazon LightSail through a cloudfront cdn and my builds will periodically fail due to a cloudfront-to-vercel error. no problems in production once live ever though. which makes me wonder if simply slapping the AWS access key/secret key pair  in would remedy this behavior 🤔 

![credentials](https://user-images.githubusercontent.com/46355797/104949948-68e08880-5985-11eb-95f5-6b3df8f0556b.png)

Vercels Serverless functions harness AWS Lambdas under the hood

**Comment by leerob** [comment]
Here's an example that works deployed to Vercel! It also uses pre-signed URLs.

https://github.com/leerob/nextjs-aws-s3

**Reply by rauchp** [reply]
I was trying to use s3.upload on server-side so I could manipulate image on server before uploading, though

**Reply by DopamineDriven** [reply]
Thanks for sharing the CDK repo link! I used your repo as a base and extended off it, still tweaking the ses and sns cdk specs for my use case but it is a huge help, really appreciate it. At any rate, I was wondering why you export `{ HelloCdkStack }` to the bin __dir for instantiation instead of instantiating the new App class in the same lib file like they do in the [docs](https://docs.aws.amazon.com/cdk/latest/guide/constructs.html#constructs_lib)?

**Reply by DopamineDriven** [reply]
never mind, that was a dumb question. I configured one from scratch using typescript and it's automatically generated in the bin __dir. for anyone curious about using typescript over js for cdk config


## Data Persistence with DynamoDB

### Sign into [AWS](https://aws.amazon.com/) as a root user

1. Navigate to [IAM](https://aws.amazon.com/iam/) from the AWS Console to create a new role
2. Select Programmatic Access for the new IAM user
3. Set Permissions -- select "Attach existing policies directly"
4. Attach the following policies to the new user
   - 4(a). AWSCloudFormationFullAccess
   - 4(b). AmazonDynamoDBFullAccess
5. With these two policies attached, click "Next: Tags" then "Next: Review"
6. Ensure that the User details and permissions are correct, then select "Create User"
7. Save the Access Key _and_ Secret Key by downloading the CSV or simply keeping the browser tab open. They will be needed to configure access to AWS from the CLI

```s
ACCESS_KEY=xxxxxxxxxxx
SECRET_KEY=xxxxxxxxxxxxxxxxxxxxxxxxxxx
```

### Configuring your local environment with AWS

- Open your terminal to your root user directory and run `aws-cli --version`
  - [Download the aws-cli](https://aws.amazon.com/cli/) if you haven't yet, then run `aws-cli --version` again to ensure it was correctly installed
- Then, run `cdk --version`

  - If this hasn't been installed, run `npm install -g aws-cdk` or `yarn global aws-cdk`

- With the CLI and CDK installed, run `aws configure` and provide the Access Key, Secret Key, default region, and output values (region likely to be `us-east-1`, output likely to be `none`)

### Open an empty project directory (git cannot be initialized yet)

- In an empty project directory, run the following command

```s
cdk init app --language typescript
```

- This initializes a new git repo and the output should look as follows

```s
$ cdk init app --language typescript
Applying project template app for typescript
# Welcome to your CDK TypeScript project!

This is a blank project for TypeScript development with CDK.

The `cdk.json` file tells the CDK Toolkit how to execute your app.

## Useful commands

 * `npm run build`   compile typescript to js
 * `npm run watch`   watch for changes and compile
 * `npm run test`    perform the jest unit tests
 * `cdk deploy`      deploy this stack to your default AWS account/region
 * `cdk diff`        compare deployed stack with current state
 * `cdk synth`       emits the synthesized CloudFormation template

Initializing a new git repository...
```

### Cdk init success

The `cdk.json` file tells the CDK Toolkit how to execute your app.

## Useful commands

- `npm run build` compile typescript to js
- `npm run watch` watch for changes and compile
- `npm run test` perform the jest unit tests
- `cdk deploy` deploy this stack to your default AWS account/region
- `cdk diff` compare deployed stack with current state
- `cdk synth` emits the synthesized CloudFormation template

### Config precedes deploy

- Before running the cdk deploy command, let's tweak some values in the tsconfig.json file
- the default tsconfig.json file generated by aws-cdk should look similar to the following

```json
// cdk-generated tsconfig.json
{
	"compilerOptions": {
		"target": "ES2018",
		"module": "commonjs",
		"lib": ["es2018"],
		"declaration": true,
		"strict": true,
		"noImplicitAny": true,
		"strictNullChecks": true,
		"noImplicitThis": true,
		"alwaysStrict": true,
		"noUnusedLocals": false,
		"noUnusedParameters": false,
		"noImplicitReturns": true,
		"noFallthroughCasesInSwitch": false,
		"inlineSourceMap": true,
		"inlineSources": true,
		"experimentalDecorators": true,
		"strictPropertyInitialization": false,
		"typeRoots": ["./node_modules/@types"]
	},
	"exclude": ["cdk.out"]
}
```

- Let's add a baseUrl as `.`, path aliases as shown below, a moduleResolution of `node`, and modify the target and lib values. The `module` value of `CommonJS` must remain as is however, otherwise cdk deploy will fail

```json
// updated tsconfig.json
{
	"compilerOptions": {
		"target": "ESNext",
		"module": "CommonJS",
		"lib": ["ESNext", "DOM", "DOM.Iterable"],
		"declaration": true,
		"strict": true,
		"noImplicitAny": true,
		"strictNullChecks": true,
		"noImplicitThis": true,
		"alwaysStrict": true,
		"moduleResolution": "node",
		"noUnusedLocals": false,
		"noUnusedParameters": false,
		"jsx": "preserve",
		"noImplicitReturns": true,
		"noFallthroughCasesInSwitch": false,
		"inlineSourceMap": true,
		"inlineSources": true,
		"experimentalDecorators": true,
		"strictPropertyInitialization": false,
		"baseUrl": ".",
		"allowJs": true,
		"skipLibCheck": true,
		"forceConsistentCasingInFileNames": true,
		"noEmit": true,
		"esModuleInterop": true,
		"resolveJsonModule": true,
		"isolatedModules": true,
		"paths": {
			"@bin/*": ["bin/*"],
			"@components/*": ["components/*"],
			"@lib/*": ["lib/*"],
			"@pages/*": ["pages/*"],
			"@scripts/*": ["scripts/*"],
			"@styles/*": ["styles/*"]
		}
	},
	"include": [
		"next-env.d.ts",
		"global.d.ts",
		"**/*.ts",
		"**/*.tsx",
		"**/*.config.js"
	],
	"exclude": ["cdk.out", ".next", "node_modules"]
}
```

## cdk-generated bin/subreddit-swr-cdk.ts

- The cdk-generated file should resemble the following

```ts
#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from '@aws-cdk/core';
import { SubredditSwrCdkStack } from '../lib/subreddit-swr-cdk-stack';

const app = new cdk.App();
new SubredditSwrCdkStack(app, 'SubredditSwrCdkStack');
```

### cdk-generated lib/subreddit-swr-cdk-stack.ts

- The cdk-generated file should resemble the following

```ts
import * as cdk from '@aws-cdk/core';

export class SubredditSwrCdkStack extends cdk.Stack {
	constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
		super(scope, id, props);

		// The code that defines your stack goes here
	}
}
```

- To configure DynamoDB, we will need to install another dependency

```git
yarn add @aws-cdk/aws-dynamodb
```

### updated lib/subreddit-swr-cdk-stack.ts

- With a mongodb table configured, the updated `subreddit-swr-cdk-stack.ts` file should look as follows

```ts
import * as cdk from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';

export class SubredditSwrCdkStack extends cdk.Stack {
	constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
		super(scope, id, props);

		new dynamodb.Table(this, 'Items', {
			partitionKey: { name: 'id', type: dynamodb.AttributeType.STRING },
			tableName: 'subredditStats',
			billingMode: dynamodb.BillingMode.PAY_PER_REQUEST
		});
	}
}
```

- Next, execute the cdk deploy script

```git
yarn cdk deploy
```

- The output should look as follows

```s
$ cdk deploy
SubredditSwrCdkStack: deploying...
SubredditSwrCdkStack: creating CloudFormation changeset...
 0/3 | 3:10:40 AM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | SubredditSwrCdkStack User Initiated
 0/3 | 3:10:45 AM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | SubredditSwrCdkStack User Initiated
 0/3 | 3:10:50 AM | CREATE_IN_PROGRESS   | AWS::DynamoDB::Table | Items (Items5C12978B)
 1/3 | 3:10:50 AM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata   | CDKMetadata/Default (CDKMetadata)
 1/3 | 3:10:50 AM | CREATE_IN_PROGRESS   | AWS::DynamoDB::Table | Items (Items5C12978B) Resource creation Initiated
 1/3 | 3:10:52 AM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata   | CDKMetadata/Default (CDKMetadata) Resource creation Initiated
 1/3 | 3:10:52 AM | CREATE_COMPLETE      | AWS::CDK::Metadata   | CDKMetadata/Default (CDKMetadata)
 3/3 | 3:11:21 AM | CREATE_COMPLETE      | AWS::DynamoDB::Table | Items (Items5C12978B)
 3/3 | 3:11:22 AM | CREATE_COMPLETE      | AWS::CloudFormation::Stack | SubredditSwrCdkStack

 ✅  SubredditSwrCdkStack

Stack ARN:
arn:aws:cloudformation:us-east-1:xxxxxxxxxxxx:stack/SubredditSwrCdkStack/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
```

### Navigate to the DynamoDB Console on AWS to inspect tables

- The desired byproduct of the programatically executed CloudFormation ChangeSet -- Success! 🎉

![DynamoDB-Table-Success](./public/dynamodb-table-created-via-cdk-deploy.jpg)

### What exactly does `cdk deploy` achieve?

- The `cdk deploy` command creates a CloudFormation ChangeSet on our behalf and abstracts away the often tedious process of managing CloudFormation directly
  - Bottom line: creates repeatable, codified, infrastructure
- Added Bonus:
  - Saves a lot of time if deploying new or modified environments frequently
  - Also useful for creating examples such as this
  - Users can choose to start from scratch or plug credentials from a similarly configured IAM user into the .env.local file of the established environment and off they go

**Reply by Manubi** [reply]
> Thanks for sharing the CDK repo link! I used your repo as a base and extended off it, still tweaking the ses and sns cdk specs for my use case but it is a huge help, really appreciate it. At any rate, I was wondering why you export `{ HelloCdkStack }` to the bin __dir for instantiation instead of instantiating the new App class in the same lib file like they do in the [docs](https://docs.aws.amazon.com/cdk/latest/guide/constructs.html#constructs_lib)?

I am trying to do the same for SNS , SMS and S3 with typescript. Can you show me how you did it? thanks!

**Comment by rauchp** [comment, marked as answer]
Fixed this. 

Before, I was just initializing [as shown here in just my next.config.js](https://vercel.com/knowledge/how-can-i-use-aws-sdk-environment-variables-on-vercel).

`aws.config.update({
  accessKeyId: process.env.AWS_ACCESS_KEY_ID_APP,
  secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY_App,
  region: 'us-east-2',
  signatureVersion: 'v4'
})
`

Turns out that's not enough. You have to be calling this inside your API route as well. I fixed this problem by adding that code block to the start of the API's function.

This was very confusing for me because it work on localhost and my own server, but not on Vercel's. But like @DopamineDriven said, Vercel is using AWS Lambda on their end for deployment, so behavior will be different on Vercel's production.
